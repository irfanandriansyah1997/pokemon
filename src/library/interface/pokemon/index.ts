export * from './pokemon-evolution.interface';
export * from './pokemon-item.interface';
export * from './pokemon-move.interface';
export * from './pokemon-species.interface';
export * from './pokemon-util.interface';
