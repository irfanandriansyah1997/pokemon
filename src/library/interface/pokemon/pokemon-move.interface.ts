import {
  IPokemonBase,
  IPokemonFlavorTextEntries
} from './pokemon-util.interface';

/**
 * Pokemon Move Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonMove {
  accuracy: number;
  damage_class: Partial<IPokemonBase>;
  effect_chance: number;
  effect_entries: { effect: string }[];
  flavor_text_entries: IPokemonFlavorTextEntries[];
  id: number;
  power: number;
  pp: number;
  target: Partial<IPokemonBase>;
  type: Partial<IPokemonBase>;
}
