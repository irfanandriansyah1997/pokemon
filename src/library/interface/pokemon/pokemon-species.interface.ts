import {
  IPokemonBase,
  IPokemonFlavorTextEntries,
  IPokemonGeneral
} from './pokemon-util.interface';

/**
 * Pokemon Species Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonSpecies {
  base_happiness: number;
  capture_rate: number;
  color: Partial<IPokemonBase>;
  egg_groups: Partial<IPokemonBase>[];
  evolution_chain: Partial<IPokemonBase>;
  flavor_text_entries: IPokemonFlavorTextEntries[];
  genera: IPokemonGeneral[];
  growth_rate: Partial<IPokemonBase>;
  habitat: Partial<IPokemonBase>;
  hatch_counter: number;
  id: number;
  is_baby: boolean;
  is_legendary: boolean;
  is_mythical: boolean;
  shape: Partial<IPokemonBase>;
}
