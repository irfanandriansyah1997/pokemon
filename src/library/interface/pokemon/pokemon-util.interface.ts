/**
 * Pokemon Base Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonBase {
  id: string;
  name: string;
  url: string;
}

/**
 * Pokemon Type Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonType {
  slot: number;
  type: Partial<IPokemonBase>;
}

/**
 * Pokemon Type Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonAbilities {
  ability: Partial<IPokemonBase>;
  is_hidden: boolean;
}

/**
 * Pokemon Stat Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonStat {
  base_stat: number;
  effort: number;
  stat: {
    name: string;
  };
}

/**
 * Pokemon Sprites Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonSprites {
  back_default: string;
  front_default: string;
}

export interface IPokemonGeneral {
  genus: string;
  language: Partial<IPokemonBase>;
}

/**
 * Pokemon Flavor Text Entries
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonFlavorTextEntries {
  flavor_text: string;
  language: Partial<IPokemonBase>;
  version?: Partial<IPokemonBase>;
  version_group?: Partial<IPokemonBase>;
}
