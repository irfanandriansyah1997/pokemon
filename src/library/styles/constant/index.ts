export * from './color.constant';
export * from './font-variable.constant';
