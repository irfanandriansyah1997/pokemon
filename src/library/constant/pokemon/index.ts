import { ITabActionItem } from '@/library/component/tab-action/interface';
import { IPokemonTab } from '@/library/features/pokemon-detail/interface';

export const MAX_POKEMON_STATS = 180;

export const POKEMON_TAB_ITEM: ITabActionItem[] = [
  {
    id: IPokemonTab.about,
    text: `About`
  },
  {
    id: IPokemonTab.stats,
    text: `Stats`
  },
  {
    id: IPokemonTab.evolution,
    text: `Evolution`
  }
];
