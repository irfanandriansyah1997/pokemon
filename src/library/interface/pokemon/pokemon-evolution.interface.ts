import { IPokemonBase } from './pokemon-util.interface';

/**
 * Pokemon Evolution Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonEvo {
  evolution_details: IPokemonEvoDetail[];
  evolves_to: IPokemonEvo[];
  species: Partial<IPokemonBase>;
}

/**
 * Pokemon Evolution Detail Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonEvoDetail {
  held_item: Partial<IPokemonBase>;
  item: Partial<IPokemonBase>;
  known_move: Partial<IPokemonBase>;
  known_move_type: Partial<IPokemonBase>;
  location: Partial<IPokemonBase>;
  min_affection: number;
  min_beauty: number;
  min_happiness: number;
  min_level: number;
  time_of_day: string;
  trigger: Partial<IPokemonBase>;
}
