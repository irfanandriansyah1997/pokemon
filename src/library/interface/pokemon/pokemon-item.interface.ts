import { IPokemonSpecies } from './pokemon-species.interface';
import {
  IPokemonAbilities,
  IPokemonBase,
  IPokemonSprites,
  IPokemonStat,
  IPokemonType
} from './pokemon-util.interface';

/**
 * Pokemon Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemon extends Partial<IPokemonOptional> {
  id: number;
  image: string;
  name: string;
}

/**
 * Pokemon Optional Data
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.30
 */
export interface IPokemonOptional {
  abilities: IPokemonAbilities[];
  base_experience: number;
  color: string;
  height: number;
  moves: { move: Partial<IPokemonBase> }[];
  nickname: string;
  pokeSpecies: IPokemonSpecies;
  species: IPokemonBase;
  sprites: IPokemonSprites;
  stats: IPokemonStat[];
  types: IPokemonType[];
  weight: number;
}
