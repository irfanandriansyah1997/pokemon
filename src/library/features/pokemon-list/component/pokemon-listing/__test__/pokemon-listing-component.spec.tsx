import { MockedProvider } from '@apollo/client/testing';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { MOCK_GRAPHQL_RESPONSE } from '@/library/__mock__/graphql-response.mock';
import MyPokemonProvider from '@/library/features/my-pokemon';
import PokemonDialog from '@/library/features/pokemon-detail/component/pokemon-dialog';
import PokemonCard from '@/library/features/pokemon-list/component/pokemon-card';
import PokemonList from '@/library/features/pokemon-list/component/pokemon-listing';
import { PokeLoadMoreButton } from '@/library/styles/pokemon.styles';

describe(`Testing Pokemon Listing`, () => {
  it(`Testing Snapshot Component`, async () => {
    const ui = mount(
      <MockedProvider mocks={MOCK_GRAPHQL_RESPONSE}>
        <MyPokemonProvider>
          <PokemonList />
        </MyPokemonProvider>
      </MockedProvider>
    );

    expect(ui.html()).toMatchSnapshot();

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    expect(ui.html()).toMatchSnapshot();
    expect(ui.find(PokemonCard).length).toBe(10);
  });

  it(`Testing Simulate Load More`, async () => {
    const ui = mount(
      <MockedProvider mocks={MOCK_GRAPHQL_RESPONSE}>
        <MyPokemonProvider>
          <PokemonList />
        </MyPokemonProvider>
      </MockedProvider>
    );

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    expect(ui.find(PokemonCard).length).toBe(10);
    expect(ui.find(PokeLoadMoreButton).length).toBe(1);

    ui.find(PokeLoadMoreButton).at(0).simulate(`click`);

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    expect(ui.find(PokemonCard).length).toBe(20);
  });

  it(`Testing Click One of PokemonCard`, async () => {
    const ui = mount(
      <MockedProvider mocks={MOCK_GRAPHQL_RESPONSE}>
        <MyPokemonProvider>
          <PokemonList />
        </MyPokemonProvider>
      </MockedProvider>
    );

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    expect(ui.find(PokemonCard).length).toBe(10);
    ui.find(PokemonCard).at(0).simulate(`click`);

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    let showDialogProps = ui.find(PokemonDialog).at(0).prop(`showDialog`);
    const onProps = ui.find(PokemonDialog).at(0).prop(`on`);
    expect(showDialogProps).toBeTruthy();
    onProps?.({
      event: `on-close`
    });

    await act(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(() => {
            ui.update();
            resolve();
          }, 1000);
        })
    );

    showDialogProps = ui.find(PokemonDialog).at(0).prop(`showDialog`);
    expect(showDialogProps).toBeFalsy();
  });
});
