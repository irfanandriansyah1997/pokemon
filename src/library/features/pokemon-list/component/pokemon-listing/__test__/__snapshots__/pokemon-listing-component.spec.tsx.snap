// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Pokemon Listing Testing Snapshot Component 1`] = `"<div class=\\"css-1m5be6d\\"><div style=\\"display: flex; flex-direction: column; margin: auto; max-width: 500px; width: 100%;\\"></div><button type=\\"submit\\" class=\\"css-1tr36ki\\"><p color=\\"white\\" font-weight=\\"500\\" class=\\"css-1wsjhlw\\">Load More</p></button></div>"`;

exports[`Testing Pokemon Listing Testing Snapshot Component 2`] = `"<div class=\\"css-1m5be6d\\"><div style=\\"display: flex; flex-direction: column; margin: auto; max-width: 500px; width: 100%;\\"><div tabindex=\\"0\\" color=\\"#ace45c\\" role=\\"button\\" class=\\"css-1p8df5j\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon bulbasaur\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#001</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">bulbasaur</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#b9414e\\" role=\\"button\\" class=\\"css-d2pu76\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon ivysaur\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#002</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">ivysaur</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#24b099\\" role=\\"button\\" class=\\"css-1mdkz8q\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon venusaur\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#003</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">venusaur</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#f49404\\" role=\\"button\\" class=\\"css-oasdhy\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon charmander\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#004</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">charmander</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#fc6104\\" role=\\"button\\" class=\\"css-1hoer6m\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon charmeleon\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/5.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#005</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">charmeleon</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#fcb414\\" role=\\"button\\" class=\\"css-l4ansr\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon charizard\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#006</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">charizard</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#e1b66e\\" role=\\"button\\" class=\\"css-k6lh0e\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon squirtle\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/7.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#007</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">squirtle</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#647cbc\\" role=\\"button\\" class=\\"css-k8gmpg\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon wartortle\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/8.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#008</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">wartortle</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#4c7cbe\\" role=\\"button\\" class=\\"css-1komheg\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon blastoise\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/9.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#009</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">blastoise</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div><div tabindex=\\"0\\" color=\\"#6cc434\\" role=\\"button\\" class=\\"css-102rfts\\"><div><img loading=\\"lazy\\" alt=\\"Pokemon caterpie\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/10.png\\"><div class=\\"pokemon-text\\"><p font-size=\\"large\\" font-weight=\\"700\\" color=\\"white\\" class=\\"css-z9ljcq\\">#010</p><p font-size=\\"normal\\" font-weight=\\"400\\" color=\\"white\\" class=\\"css-1r2ireq\\">caterpie</p></div><div class=\\"poke-ball css-adl4sn\\"></div></div></div></div><button type=\\"submit\\" class=\\"css-1k960ei\\"><p color=\\"white\\" font-weight=\\"500\\" class=\\"css-1wsjhlw\\">Load More</p></button><div color=\\"#ddd\\" class=\\"css-1b90bes\\"></div><div class=\\"css-1b90bes\\"></div><button class=\\"css-j4r94v\\"><div class=\\"poke-ball css-adl4sn\\"></div></button><div color=\\"rgba(0,0,0,0.25)\\" class=\\"css-133wia2\\"></div><div class=\\"css-swco5x\\"><div class=\\"css-ulx48j\\"><div class=\\"css-8prdx5\\"><p font-size=\\"normal\\" color=\\"primary\\" font-weight=\\"500\\" class=\\"css-nftjgg\\">Catch Pokemon Succeess!</p><div class=\\"css-136ng3n\\"><p font-size=\\"medium\\" class=\\"css-1109ac7\\">×</p></div></div><div class=\\"css-we45xm\\"><input type=\\"text\\" placeholder=\\"Give a pokemon name\\" class=\\"css-5e2eqg\\" value=\\"\\"><p color=\\"red\\" class=\\"css-1gkutas\\"></p></div><div class=\\"css-1m7yzt4\\"><button type=\\"button\\" disabled=\\"\\" class=\\"css-jsg0f6\\">Save</button><button type=\\"button\\" class=\\"css-bbsgag\\">Cancel</button></div></div></div><div color=\\"rgba(0,0,0,0.25)\\" class=\\"css-133wia2\\"></div><div class=\\"css-swco5x\\"><div class=\\"css-ulx48j\\"><div class=\\"css-8prdx5\\"><p font-size=\\"normal\\" color=\\"primary\\" font-weight=\\"500\\" class=\\"css-nftjgg\\">Failed</p><div class=\\"css-136ng3n\\"><p font-size=\\"medium\\" class=\\"css-1109ac7\\">×</p></div></div><div class=\\"css-we45xm\\"><p class=\\"css-5jbmio\\">It ran away...</p></div></div></div><div color=\\"rgba(0,0,0,0.25)\\" class=\\"css-133wia2\\"></div><div class=\\"css-swco5x\\"><div class=\\"css-ulx48j\\"><div class=\\"css-8prdx5\\"><p font-size=\\"normal\\" color=\\"primary\\" font-weight=\\"500\\" class=\\"css-nftjgg\\">Failed</p><div class=\\"css-136ng3n\\"><p font-size=\\"medium\\" class=\\"css-1109ac7\\">×</p></div></div><div class=\\"css-we45xm\\"><p class=\\"css-5jbmio\\">Are you sure delete pokemon <b style=\\"color: rgb(26, 26, 26); font-weight: 500;\\"></b> ?</p></div><div class=\\"css-1m7yzt4\\"><button type=\\"button\\" class=\\"css-1q06zjp\\">Cancel</button><button type=\\"button\\" class=\\"css-bbsgag\\">Delete</button></div></div></div></div>"`;
